crypto/rand/rand_vxworks.c

```diff
@@ -0,0 +1,171 @@
+ */
+ 
+#ifndef _GNU_SOURCE
+# define _GNU_SOURCE
+#endif
+
+#if defined(OPENSSL_SYS_VXWORKS)
+
+#include <openssl/rand.h>
+#include "rand_lcl.h"
+#include "internal/rand_int.h"
+#include "internal/cryptlib.h"
+
+# if defined(OPENSSL_RAND_SEED_OS)    
+#  include <randomNumGen.h>
+# endif
+
+/* Macro to convert two thirty two bit values into a sixty four bit one */
+# define TWO32TO64(a, b) ((((uint64_t)(a)) << 32) + (b))
+
+
+static uint64_t get_time_stamp(void)
+{
+    struct timespec ts;
+
+    if (clock_gettime(CLOCK_REALTIME, &ts) == 0)
+        return TWO32TO64(ts.tv_sec, ts.tv_nsec);
+	return time(NULL);
+}
+
+static uint64_t get_timer_bits(void)
+{
+    uint64_t res = OPENSSL_rdtsc();
+
+    if (res != 0)
+        return res;
+
+    {
+        struct timespec ts;
+
+        if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0)
+            return TWO32TO64(ts.tv_sec, ts.tv_nsec);
+    }
+    return time(NULL);
+}
+
+/* 
+ *	empty implementation 
+ *  vxworks does not need to init/cleanup or keep open the random lib
+ */
+int rand_pool_init(void)
+{
+    return 1;
+}
+
+void rand_pool_cleanup(void)
+{
+}
+
+void rand_pool_keep_random_devices_open(int keep)
+{
+}
+
+int rand_pool_add_additional_data(RAND_POOL *pool)
+{
+    struct {
+        CRYPTO_THREAD_ID tid;
+        uint64_t time;
+    } data = { 0 };
+
+    /*
+     * Add some noise from the thread id and a high resolution timer.
+     * The thread id adds a little randomness if the drbg is accessed
+     * concurrently (which is the case for the <master> drbg).
+     */
+    data.tid = CRYPTO_THREAD_get_current_id();
+    data.time = get_timer_bits();
+
+    return rand_pool_add(pool, (unsigned char *)&data, sizeof(data), 0);
+}
+
+int rand_pool_add_nonce_data(RAND_POOL *pool)
+{
+    struct {
+        pid_t pid;
+        CRYPTO_THREAD_ID tid;
+        uint64_t time;
+    } data = { 0 };
+
+    /*
+     * Add process id, thread id, and a high resolution timestamp to
+     * ensure that the nonce is unique whith high probability for
+     * different process instances.
+     */
+    data.pid = getpid();
+    data.tid = CRYPTO_THREAD_get_current_id();
+
+    data.time = get_time_stamp();
+
+    return rand_pool_add(pool, (unsigned char *)&data, sizeof(data), 0);
+}
+# if defined(OPENSSL_RAND_SEED_NONE)
+size_t rand_pool_acquire_entropy(RAND_POOL *pool)
+{
+    short int code;
+    int i, k;
+    size_t bytes_needed;
+    struct timespec ts;
+    unsigned char v;
+    long long duration;
+    extern void s$sleep2(long long *_duration, short int *_code);
+
+    bytes_needed = rand_pool_bytes_needed(pool, 4 /*entropy_factor*/);
+
+    for (i = 0; i < bytes_needed; i++) {
+        /*
+         * burn some cpu; hope for interrupts, cache collisions, bus
+         * interference, etc.
+         */
+        for (k = 0; k < 99; k++)
+            ts.tv_nsec = random();
+
+        /* sleep for 1/65536 of a second (15 us).  */
+        duration = 1;
+        s$sleep2(&duration, &code);

```

Comment:
i kind of lost your original suggestion for gathering entropy w/o the randlib, could you suggest somethign for this entire function? (we are using the randlib so i just copied over what i found in the rand_unix.c)

Reply:
Have you tested your code on a vxWorks target machine?
I mainly have my wisdom from looking at source code, which is available to windriver
customers. But I am not set up to do much testing especially on a real target.

I _believe_ it would be better entropy to use busy loop here, that
counts how often clock_gettime returns the same value like this one:
```
    do {
      clock_gettime(CLOCK_MONOTONIC, &t2);
      picosec++;
    } while (t1.tv_sec == t2.tv_sec && t1.tv_nsec == t2.tv_nsec);
```

what do you think?

Reply:
I mean do this instead of the sleep2 code.

Reply:
overall this function looks good.

Reply:
just went back to my old vxworks 6.9 install, do you think we could consider using prng.h's functions to add entropy and skip the clock stuff here entirely?